manpages/mrc2tif.man
.na
.nh
.TH mrc2tif 1 2.00 BL3DEMC
.SH NAME
mrc2tif \- Create TIFF, JPEG, or PNG files from an MRC image file.
.SH SYNOPSIS
mrc2tif  [options]  mrc_input_file  tiff_file_root_name
.SH DESCRIPTION
mrc2tif usually creates a series of TIFF files with the prefix \fItiff_file_root_name\fR
and with the suffix .nnn.tif, where nnn is the z number.  Byte, signed and
unsigned 16-bit integer, float, 
and RGB color files (modes 0, 1, 6, 2, 16) can be converted.  There are
options to create JPEG or PNG files instead of TIFF.  The data will be
written to a TIFF file in the same mode as in the input file by default, but
there are options that allow the integer and float modes to be scaled to
bytes; the same options allow scaling of byte and RGB values.  If the input
file has only one section, then the output file will simply be the 
\fItiff_file_root_name\fR with no suffix added.  
.P
When the output mode is integer or float, the TIFF file will contain the
minimum and maximum value for the data in the file.  When a TIFF stack is
created, each
image will have the same overall minimum and maximum, rather than values
specific to that image.  This way, 3dmod will be able to read
in the data with proper scaling.  However, Photoshop pays no attention to
these values when displaying integer data, so when the integer data have a
relatively small range, the image will initially appear black or very dark.
The simplest solution for this problem is to apply the scaling options
described next.
.P
When either the
.B -S
or the
.B -C
option is entered, integer and floating point data will be scaled to bytes and
any values outside the range of 0 to 255 will be truncated.  If the input data
is already in bytes (either byte data or RGB data), these options will cause
the data to be scaled, and values outside the range truncated.  To have the
range of the input data mapped into the full range of byte data, just enter
"-C 0,255".  To have a subset of the input range mapped to 0 to 255, either
enter the limits of the subset of the range directly with the 
.B -S
option, or adjust the contrast in 3dmod(1) to give the desired expansion of
the dynamic range for that subset of the image data, and enter the contrast
settings withs the 
.B -C
option.
.P Output to JPEG or PNG files is selected by entering the
.B -j
or the
.B -p
option.  There are some restrictions in these cases.  An output stack cannot
be created, only a series of single-image files.  The type of compression
cannot be specified, but the quality or degree of compression can.  Tiling
cannot be specified.  Only byte and RGB output is available.  Integer and
floating point files are converted to byte based either on the minimum and maximum
densities in the input file or on the values entered with 
.B -S
or
.B -C.
In addition to the fact that JPEG images cannot be bigger than 65535 pixels in
X or Y, problems may be encountered with images much bigger than 1 gigapixel.
.P
Since IMOD is distributed with version 4 of the TIFF library, this program
will write files bigger than 4 GB in the BigTIFF format, and continue to write
smaller files in classical TIFF format.  
.SH OPTIONS
.TP
.B -s
Create a single output file, a TIFF stack with all of
the sections in the input file.
.TP
.B -j
Produce JPEG files instead of TIFF files.  See above for the restrictions with
this kind of output.
.TP
.B -p
Produce PNG files instead of TIFF files.  See above for the restrictions.
.TP
.B -c \fIvalue\fR
Compress the data in TIFF files with the type of compression indicated.  The
most useful compression types are LZW (value 5), Zip or Deflate (value 8), and
JPEG (value 7).  For these types, the value can be given as a text string:
"lzw", "zip", or "jpeg".  These are available on all operating systems when
running with the TIFF libraries distributed with IMOD.  On Linux, see 
/usr/include/tiff.h for other numeric
values that can be tried; some of these will work, depending on the data type.
Also note that JPEG compression is not available with 16-bit data.
.TP
.B -q \fIvalue\fR
Set the quality for JPEG or ZIP compression (ZIP compression is used in PNG
output).  For JPEG, the quality ranges from
0 to 100 with a default of 75; higher numbers preserve more information but
with less compression.  For ZIP/PNG, which is a lossless compression, the quality
factor ranges from 1 to 9 with a default of 6 and specifies a tradeoff between
speed and amount of compression.  Higher numbers giving somewhat more
compression for more computational time.  Lower numbers give somewhat less
compression with potentially much less computational time.
.TP
.B -S \fImin,min\fR
Specify an initial intensity scaling that will scale "min" to 0 and "max" to
255, just as is achieved with the 
.B -s
options to 3dmod(1) and mrcbyte(1).  
.TP
.B -C \fIblack,white\fR
Specify intensity scaling to match the scaling achieved with the given
black and white contrast settings in 3dmod(1), with the -c option to
mrcbyte(1), or the "-contrast" option to newstack.  If both
.B -S
and
.B -C
are entered, the final scaling will be the product of these two scalings.
If
.B -S
is entered alone, the black/white values are 0/255 (no scaling for contrast).
If
.B -C
is entered alone, the initial scaling will be from the minimum in the input
file to 0, and from the maximum in the file to 255.  If you have integer data
and you want to convert it to bytes, you will generally want to apply some
scaling with black/white values to preserve the dynamic range of the data of
interest.
.TP
.B -z \fImin,max\fR
 Starting and ending Z slices in file to output, numbered from 0.
.TO
.B -i \fIvalue\fR
Initial file number.  The default is the number of the first Z slice output,
where Z is numbered from 0.
.B -r \fIvalue\fR
Set the resolution setting of the output file in dots per inch to the given
value.
.B -P
Use the pixel spacing in the MRC file header to set the resolution in
the output file.  The resolution will be set in dots per inch and will
typically be a very large number, but this value allows the pixel spacing to
be recovered if the file is converted back with Tif2mrc(1).
.TP
.B -T \fIxsize,ysize\fR or \fIsize\fR
Organize output data in tiles instead of strips, which can allow faster access
to subareas of very large images.  Enter one number for the tile size in both
dimensions, or two numbers to specify a different size in X and Y.  The TIFF
library requires that all tiles be the same size, which means that the image
file will be padded so that tiles on the bottom and right sides have
the same sizes as the rest.  The program will pick an actual tile size near
the entered number that is a multiple of 16 and that minimizes the amount of
padding required.
.TP
.B -t \fIvalue\fR
Set the criterion for processing each image in strips to the given value in
megabytes (fractional values are allowed).  Whole images will be 
read in, processed, and saved as a unit if they are less than this criterion
(default 100 MB).  Above the criterion, images will be read, processed, and
saved in strips, with much lower memory requirements.
.TP
.B -o
Use old IMOD code instead of libtiff for writing the files (for testing
purposes).
.SH AUTHORS
Jim Kremer, David Mastronarde
.SH SEE ALSO
clip, newstack
.SH BUGS
Complex and long integer data modes are not supported.

Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

mrc/mrc2tif.c
/*
 *  mrc2tif.c -- Convert mrc files to TIFF files.
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 *  Log at end
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "mrcfiles.h"
#include "b3dtiff.h"
#include "b3dutil.h"
#include "parse_params.h"
static FILE *openEitherWay(ImodImageFile *iifile, char *iname, char *progname,
                           int oldcode);

int main(int argc, char *argv[])
{
  FILE *fin;
  MrcHeader hdata;
  FILE *fpTiff = NULL;
  Islice slice;
  int xsize, ysize, zsize, psize, filenum, outPsize, numChunks, chunk;
  int lines, linesDone, version, linesPerChunk, doChunks = 0;
  size_t xysize, allocSize;
  int i, j, slmode, z, iarg = 1, stack = 0, resolution = 0, initialNum = -1;
  int oldcode = 0, convert = 0, usePixel = 0;
  int compression = 1, black = 0, white = 255, zmin = -1, zmax = -1;
  int quality = -1;
  b3dUInt32 ifdOffset = 0, dataOffset = 0;
  float chunkCriterion = 100.;
  float savecrit, dmin, dmax, smin =0., smax = 0.;
  float val[3];
  float scale, offset, xscale, yscale, zscale;
  char prefix[100];
  IloadInfo li;
  ImodImageFile *iifile = iiNew();
  char *iname;
  unsigned char *buf;
  char *progname = imodProgName(argv[0]);
#define NUM_LEGAL 12
  int legalComp[NUM_LEGAL] = {1, 2, 3, 4, 5, 7, 8, 32771, 32773, 32946, 32845,
                              32845};

  sprintf(prefix, "\nERROR: %s - ", progname);
  setExitPrefix(prefix);

  for (iarg = 1; iarg < argc - 1 ; iarg++){
    if (argv[iarg][0] == '-'){
      switch (argv[iarg][1]){
      case 's':
        stack = 1;
        break;

      case 'o':
        oldcode = 1;
        break;

      case 'c':
        iarg++;
        if (!strcmp(argv[iarg], "zip"))
          compression = 8;
        else if (!strcmp(argv[iarg], "jpeg"))
          compression = 7;
        else if (!strcmp(argv[iarg], "lzw"))
          compression = 5;
        else
          compression = atoi(argv[iarg]);
        for (i = 0; i < NUM_LEGAL; i++)
          if (compression == legalComp[i])
            break;
        if (i == NUM_LEGAL)
          exitError("Compression value %d not allowed", compression);
        break;

      case 'r':
        resolution = atoi(argv[++iarg]);
        break;

      case 'P':
        usePixel = 1;
        break;

      case 'S':
        sscanf(argv[++iarg], "%f%*c%f", &smin, &smax);
        convert = 1;
        break;

      case 'C':
        sscanf(argv[++iarg], "%d%*c%d", &black, &white);
        convert = 1;
        break;

      case 'z':
        sscanf(argv[++iarg], "%d%*c%d", &zmin, &zmax);
        break;

      case 'i':
        initialNum = atoi(argv[++iarg]);
        break;

      case 'q':
        quality = atoi(argv[++iarg]);
        break;

      case 't':
        chunkCriterion = atof(argv[++iarg]);
        break;

      default:
        break;
      }
      
    }
    else
      break;
       
  }

  if (oldcode && compression != 1)
    exitError("Compression not available with old writing code");
  if (oldcode && (resolution > 0 || usePixel))
    exitError("Resolution setting is not available with old writing code");
  if (resolution > 0 && usePixel)
    exitError("You cannot enter both -r and -P");

  if (argc - iarg != 2){
    printf("%s version %s \n", progname, VERSION_NAME);
    imodCopyright();
    printf("%s [options] <mrc file> <tiff name/root>\n\n", progname);
    printf(" Without -s, a series of tiff files will be created "
            "with the\n prefix [tiff root name] and with the suffix nnn.tif, "
           "where nnn is the z number. \n  Options:\n");
    printf("    -s         Stack all images in the mrc file into a single tiff"
           " file\n");
    printf("    -c val     Compress data; val can be lzw, zip, jpeg, or "
           "numbers defined\n\t\t in libtiff\n");
    printf("    -q #       Quality for jpeg compression (0-100) or for zip "
           "compression (1-9)\n");
    printf("    -S min,max Initial scaling limits for conversion to bytes\n");
    printf("    -C b,w     Contrast black/white values for conversion to "
           "bytes\n");
    printf("    -z min,max Starting and ending Z (from 0) to output\n");
    printf("    -i #       Initial file number (default is starting Z)\n");
    printf("    -r #       Resolution setting in dots per inch\n");
    printf("    -P         Use pixel spacing in MRC header for resolution setting\n");
    printf("    -t #       Criterion image size in megabytes for processing "
           "file in strips\n");
    printf("    -o         Write file with old IMOD code instead of libtiff"
           "\n");
    exit(1);
  }

  if (NULL == (fin = fopen(argv[iarg], "rb")))
    exitError("Couldn't open %s", argv[iarg]);

  if (mrc_head_read(fin, &hdata))
    exitError("Can't Read Input Header from %s", argv[iarg]);
  iarg++;
  if (usePixel) {
    mrc_get_scale(&hdata, &xscale, &yscale, &zscale);
    resolution = 2.54e8 / xscale;
  }

  if (zmin == -1 && zmax == -1) {
    zmin = 0;
    zmax = hdata.nz - 1;
  } else if (zmin < 0 || zmax >= hdata.nz || zmin > zmax)
    exitError("zmin,zmax values are reversed or out of the range 0 to %d\n",
              hdata.nz - 1);
  if (initialNum < 0)
    filenum = zmin;
  else
    filenum = initialNum;

  iifile->format = IIFORMAT_LUMINANCE;
  iifile->file = IIFILE_TIFF;
  iifile->type = IITYPE_UBYTE;
  iifile->amin = iifile->amax = 0;

  switch(hdata.mode){
      
  case MRC_MODE_BYTE:
    psize = 1;
    break;
  case MRC_MODE_SHORT:
    psize = 2;
    iifile->type = IITYPE_SHORT;
    break;
  case MRC_MODE_USHORT:
    psize = 2;
    iifile->type = IITYPE_USHORT;
    break;
  case MRC_MODE_RGB:
    psize = 3;
    iifile->format = IIFORMAT_RGB;
    break;
  case MRC_MODE_FLOAT:
    psize = 4;
    iifile->type = IITYPE_FLOAT;
    break;
  default:
    exitError("Data mode %d not supported.", hdata.mode);
    break;
  }
  xsize = hdata.nx;
  ysize = hdata.ny;
  zsize = hdata.nz;
  xysize = (size_t)xsize * (size_t)ysize;
  allocSize = xysize * psize;
  iifile->nx = xsize;
  iifile->ny = ysize;
  iifile->nz = 1;
  li.xmin = 0;
  li.xmax = xsize - 1;
  li.ymin = 0;
  li.ymax = ysize - 1;
  dmin = hdata.amin;
  dmax = hdata.amax;
  slmode = sliceModeIfReal(hdata.mode);
  outPsize = psize;
  if (convert) {
    mrcContrastScaling(&hdata, smin, smax, black, white, MRC_RAMP_LIN, &scale, 
                       &offset);
    if (slmode > 0) {
      iifile->type = IITYPE_UBYTE;
      outPsize = 1;
    }
  }
  version = tiffVersion(&j);
  /* printf("Version %d.%d\n", version,j); */
  if (version < 4) {
    savecrit =  4.292e9;
    if (!version || oldcode)
      savecrit =  2.146e9;
    if ((double)xysize * outPsize >  savecrit)
      exitError("The image is too large in X/Y to save with TIFF version"
                " %d.%d", version, j);
    if (stack && (zmax + 1 - zmin) * xysize * (double)outPsize >  savecrit)
      exitError("The volume is too large to save in a stack with TIFF version"
                " %d.%d", version, j);
  }

  if (version > 0 && !oldcode && 
      ((double)xsize * ysize) * psize > chunkCriterion * 1024. * 1024.)
    doChunks = 1;

  iname = (char *)malloc(strlen(argv[iarg]) + 20);
  if (!iname)
    exitError("Failed to allocate memory for name");

  /* Open new file for stack */
  if (stack) {
    sprintf(iname, "%s", argv[iarg]);
    iifile->nz = zmax + 1 - zmin;
    fpTiff = openEitherWay(iifile, iname, progname, oldcode);
  }    

  printf("Writing TIFF images. ");
  for (z = zmin; z <= zmax; z++){
    int tiferr;
    printf(".");
    fflush(stdout);

    /* Get filename and open new file for non-stack */
    if (!stack) {
      sprintf(iname, "%s.%3.3d.tif", argv[iarg], filenum++);
      if (zsize == 1)
        sprintf(iname, "%s", argv[iarg]);
      if (z > zmin)
        free(iifile->filename);
      fpTiff = openEitherWay(iifile, iname, progname, oldcode);
      ifdOffset = 0;
      dataOffset = 0;
    }
    if (!stack && slmode > 0 && !convert) {
      iifile->amin = 1.e30;
      iifile->amax = -1.e30;
    }

    /* If doing chunks, set up chunk loop */
    numChunks = 1;
    if (doChunks) {
      tiffWriteSetup(iifile, compression, 0, resolution, quality, 
                     &linesPerChunk, &numChunks);
      allocSize = xsize * linesPerChunk * psize;
      linesDone = 0;
    }
    
    for (chunk = 0; chunk < numChunks; chunk++) {

      /* Allocate memory first time or every time */
      if ((!chunk && z == zmin) || (convert && slmode > 0))
        buf = (unsigned char *)malloc(allocSize);
      if (!buf)
        exitError("Failed to allocate memory for slice");

      lines = ysize;
      if (doChunks) {
        lines = B3DMIN(linesPerChunk, ysize - linesDone);
        li.ymin = ysize - (linesDone + lines);
        li.ymax = li.ymin + lines - 1;
        linesDone += lines;
      }

      if (mrcReadZ(&hdata, &li, buf, z)) {
        perror("mrc2tif ");
        exitError("Reading section %d", z);
      }
      sliceInit(&slice, xsize, lines, hdata.mode, buf);
      
      /* Scale the slice and convert it to bytes */
      if (convert) {
        for (j = 0; j < lines; j++) {
          for (i = 0; i < xsize; i++) {
            sliceGetVal(&slice, i, j, val);
            val[0] = val[0] * scale + offset;
            val[0] = B3DMAX(0., B3DMIN(255., val[0]));
            if (psize == 3) {
              val[1] = val[1] * scale + offset;
              val[1] = B3DMAX(0., B3DMIN(255., val[1]));
              val[2] = val[2] * scale + offset;
              val[2] = B3DMAX(0., B3DMIN(255., val[2]));
            }
            slicePutVal(&slice, i, j, val);
          }
        }
        if (slmode > 0 && sliceNewMode(&slice, SLICE_MODE_BYTE))
          exitError("Converting slice %d to bytes", z);
        buf = slice.data.b;
      }
    
      /* Get min/max for int/float images, those are the only ones the write
         routine will put out min/max for */
      if (!stack && slmode > 0 && !convert) {
        sliceMMM(&slice);
        iifile->amin = B3DMIN(iifile->amin, slice.min);
        iifile->amax = B3DMAX(iifile->amax, slice.max);
      }

      if (fpTiff)
        tiferr = tiff_write_image(fpTiff, xsize, ysize, hdata.mode, buf, 
                                  &ifdOffset, &dataOffset, dmin, dmax);
      else if (!doChunks)
        tiferr = tiffWriteSection(iifile, buf, compression, 0, resolution,
                                  quality);
      else
        tiferr = tiffWriteStrip(iifile, chunk, buf);
      if (tiferr){
        exitError("Error (%d) writing section %d to %s", tiferr, z, iname);
      }
      if (convert && slmode > 0)
        free(slice.data.b);
    }
    if (doChunks)
      tiffWriteFinish(iifile);
    
    if (!stack) {
      if (fpTiff) {
        fclose(fpTiff);
        fpTiff = NULL;
      } else
        iiClose(iifile);
    }
  }

  printf("\r\n");
  if (stack) {
    if (fpTiff)
      fclose(fpTiff);
    else
      iiClose(iifile);
  }
  fclose(fin);
  exit(0);
}

static FILE *openEitherWay(ImodImageFile *iifile, char *iname, char *progname,
                           int oldcode)
{
  static int warned = 0;
  FILE *fpTiff = NULL;
  iifile->filename = strdup(iname);
  
  if (oldcode || tiffOpenNew(iifile)) {
    fpTiff = fopen(iname, "wb");
    if (!fpTiff){
      perror("mrc2tif system message");
      exitError("Opening %s", iname);
    } else if (!oldcode && !warned) {
      warned = 1;
      printf("\nWARNING: %s - Not writing with libtiff, compression not "
             "available", progname);
    }
  }
  return fpTiff;
}

/*

$Log$
Revision 3.15  2010/12/18 18:46:20  mast
Now it will work with up to 4 GB with version 3 libtiff, and above 4 GB
for libtiff 4.  Reads files in chunks corresponding to output file strips
above a certain size.

Revision 3.14  2010/12/17 06:20:51  mast
Maybe it will work with > 2 GB of data

Revision 3.13  2010/12/15 06:21:24  mast
Added options for setting resolution, controlling scaling, setting file
number and doing a subset in Z

Revision 3.12  2009/11/27 16:38:09  mast
Added include to fix warnings

Revision 3.11  2009/04/01 00:20:34  mast
Call new writing routine with libtiff and add compression option

Revision 3.10  2008/05/24 14:52:45  mast
Fixed string length allocation

Revision 3.9  2008/05/23 23:03:31  mast
Added string include

Revision 3.8  2008/05/23 22:57:17  mast
Added float support, option for single stack, and standardized errors

Revision 3.7  2006/06/19 19:29:23  mast
Added ability to write unsigned ints from mode 6

Revision 3.6  2006/02/19 17:53:53  mast
Allocate an array only big enough for the data type, test on array 
allocation, and issue standard error messages (but still to stderr)

Revision 3.5  2005/02/11 01:42:34  mast
Warning cleanup: implicit declarations, main return type, parentheses, etc.

Revision 3.4  2004/11/05 18:53:10  mast
Include local files with quotes, not brackets

Revision 3.3  2004/07/07 19:25:30  mast
Changed exit(-1) to exit(3) for Cygwin

Revision 3.2  2003/10/24 02:28:42  mast
strip directory from program name and/or use routine to make backup file

Revision 3.1  2002/11/05 23:48:02  mast
Changed to call imodCopyright

*/

